pkg/operator/watchdog/cmd.go

```diff
@@ -0,0 +1,247 @@
+	o := NewFileWatcherOptions()
+
+	cmd := &cobra.Command{
+		Use:   "file-watcher-watchdog",
+		Short: "Watch files on the disk and kill the specified process on change",
+		Run: func(cmd *cobra.Command, args []string) {
+			klog.V(1).Info(cmd.Flags())
+			klog.V(1).Info(spew.Sdump(o))
+
+			ctx := context.TODO()
+
+			if err := o.Complete(); err != nil {
+				klog.Fatal(err)
+			}
+			if err := o.Validate(); err != nil {
+				klog.Fatal(err)
+			}
+
+			if err := o.Run(ctx); err != nil {
+				klog.Fatal(err)
+			}
+		},
+	}
+
+	o.AddFlags(cmd.Flags())
+
+	return cmd
+}
+
+func (o *FileWatcherOptions) AddFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&o.ProcessName, "process-name", "", "name of the process to send signal to on file change (eg. 'hyperkube').")
+	fs.StringSliceVar(&o.Files, "files", o.Files, "comma separated list of file names to monitor for changes")
+	fs.StringVar(&o.KubeConfig, "kubeconfig", o.KubeConfig, "kubeconfig file or empty")
+	fs.StringVar(&o.Namespace, "namespace", o.Namespace, "namespace to report the watchdog events")
+	fs.DurationVar(&o.Interval, "interval", 5*time.Second, "interval specify how aggressive the file checks should be")
+}
+
+func (o *FileWatcherOptions) Complete() error {
+	clientConfig, err := client.GetKubeConfigOrInClusterConfig(o.KubeConfig, nil)
+	if err != nil {
+		return err
+	}
+	kubeClient, err := kubernetes.NewForConfig(clientConfig)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(context.TODO(), 10*time.Second)
+	defer cancel()
+
+	// Get event recorder.
+	// Retry on connection errors, but don't error out, instead fallback to the namespace.
+	var eventTarget *v1.ObjectReference
+	err = retry.RetryOnConnectionErrors(ctx, func(context.Context) (bool, error) {
+		var clientErr error
+		eventTarget, clientErr = events.GetControllerReferenceForCurrentPod(kubeClient, o.Namespace, nil)
+		if clientErr != nil {
+			return false, clientErr
+		}
+		return true, nil
+	})
+	if err != nil {
+		klog.Warningf("unable to get owner reference (falling back to namespace): %v", err)
+	}
+	o.recorder = events.NewRecorder(kubeClient.CoreV1().Events(o.Namespace), "file-change-watchdog", eventTarget)
+
+	return nil
+}
+
+func (o *FileWatcherOptions) Validate() error {
+	if len(o.ProcessName) == 0 {
+		return fmt.Errorf("process name must be specified")
+	}
+	if len(o.Files) == 0 {
+		return fmt.Errorf("at least one file to observe must be specified")
+	}
+	if len(o.Namespace) == 0 && len(os.Getenv("POD_NAMESPACE")) == 0 {
+		return fmt.Errorf("either namespace flag or POD_NAMESPACE environment variable must be specified")
+	}
+	return nil
+}
+
+// startPidObserver runs a loop that observe changes to the PID of the process we send signals after change is detected.
+// When the PID is observed for the first time, the pidObservedCh is closed.
+// When the PID change is observed (after we got initial PID), the shutdown is called (to signal watchdog to restart).
```

Comment:
why do we restart? Isn't there a chance this way that the watchdog misses file changes?

* operand starts
* sidecar start
* secret changes
* sidecar kills operand
* sidecar terminates, operand terminates
* operand start
* secret changes again
* sidecar start
=> boom. Latest change not noticed.

Reply:
tl/dr: I don't like the sidecar to restart at all. It shouldn't.

Reply:
I'm more worried about the watchdog to be "stucked" or reading the files from wrong directory on PID change. If we don't want to restart, it means we have to make the logic in the `Run()` more complex and less error prone.

Why you hate restarts?

Reply:
restarts mean races. I don't like races.

Reply:
ok, lets go with no restart, I updated the code to run in loop, contexts are fun.